backend/api/src/realms.rs

```diff
@@ -25,73 +38,126 @@ impl Realm {
         // We store the nodes of the realm tree in a hash map
         // accessible by the database ID
-        let realms = Self::raw_from_db(db).await?
+        let mut realms = db.get().await?
+            .query_raw(
+                "select id, name, parent, path_segment from realms",
+                std::iter::empty(),
+            ).await?
+            .map_ok(|row| {
+                let key = row.get_key(0);
+                Realm {
+                    key,
+                    name: row.get(1),
+                    parent_key: if key == 0 { None } else { Some(row.get_key(2)) },
+                    path_segment: row.get(3),
+                    child_keys: vec![],
+                }
+            })
             .map_ok(|realm| (realm.key, realm))
             .try_collect::<HashMap<_, _>>().await?;
 
         // With this, and the `parent` member of the `Realm`,
         // we already have quick access to the data of a realm's parent.
-        // To also get to the children quickly we maintain another map.
-        let mut children = <HashMap<_, Vec<_>>>::new();
+        // To also get to the children quickly we maintain a corresponding list
+        // for each realm
+        let keys = realms.values()
+            .filter_map(
+                |realm| realm.parent_key.map(
+                    |parent_key| (realm.key, parent_key)
+                )
+            )
+            .collect::<Vec<_>>();
+        for (key, parent_key) in keys {
+            let parent = realms.get_mut(&parent_key).context("invalid parent")?;
+            parent.child_keys.push(key);
+        }
+
+        // After this point, we should know the tree structure to be valid.
+        // That is, we can now safely panic if we can't find things in our maps/lists;
+        // that's totally a bug in this code, then, not an inconsistency in the db.
+
+        // We also need a map from the full path to the proper realm,
+        // and conversely, we want to cache the full path inside the realm.
+        let from_path = index_by_path(&mut realms);
+
+        fn index_by_path(
+            realms: &mut HashMap<u64, Realm>,
+        ) -> HashMap<String, u64> {
 
-        for Realm { key, parent_key, .. } in realms.values() {
-            if key != parent_key {
-                children.entry(*parent_key).or_default().push(*key);
+            let mut index = HashMap::new();
+
+            fill_index_recursively(0, "", realms, &mut index);
+
+            fn fill_index_recursively(
+                parent: u64,
+                full_path: &str,
+                realms: &HashMap<u64, Realm>,
+                index: &mut HashMap<String, u64>,
+            ) {
+                index.insert(full_path.to_owned(), parent);
+
+                let parent = &realms[&parent];
+                for child_key in &parent.child_keys {
+                    let child = &realms[&child_key];
+                    let path = [full_path, &child.path_segment].join("/");
```

Comment:
I think you would rather `format!("{}/{}", ...)` here.

Reply:
Yeah I thought about it, and that's what I had, before, but now I like this more. :sweat_smile:

The reasoning being: It's semantically closer to what we want to do.

Plus it's probably more efficient, right? I mean it should have to do less work. :thinking: But also somethingsomething premature optimization of course.

Reply:
Why would that be less work? And well, I don't care, you can also keep it like this.

Reply:
`format!` has to parse the format string for example. Like, it's a more generic facility, it has to have to do more work, no?

Reply:
Oh and also this mirrors what `Realm::path` does more closely. :wink:

Reply:
The format string is parsed at compile time (of course, :wink:). And formatting a string is also just a clone. However, I just benchmarked and you are right:

```
test bench_format ... bench:          79 ns/iter (+/- 0)
test bench_join   ... bench:          22 ns/iter (+/- 0)
```

Tested with "lectures/2020" and "physics".